tests/backend/util/test_devices.py

```diff
@@ -0,0 +1,63 @@
+"""
+Test abstract device class.
+"""
+
+import pytest
+import torch
+
+from invokeai.app.services.config import get_config
+from invokeai.backend.util.devices import TorchDeviceSelect
+
+devices = ["cpu", "cuda:0", "cuda:1", "mps"]
+device_types = [
+    ("cpu", torch.float32),
+    ("cuda:0", torch.float16 if TorchDeviceSelect.has_cuda else torch.float32),
+    ("mps", torch.float16 if TorchDeviceSelect.has_mps else torch.float32),
```

Comment:
We should really be mocking the `has_cuda` and `has_mps` checks throughout this file so that we exercise all cases regardless of which system we run on.

Or, as a slightly less optimal solution (but probably simpler) we could use `pytest.mark.skipif` for tests that can't run on the current HW.

Reply:
This is done.